Add option to clear all notes in SlicerT

bratpeki opened the pull request:
Adds a button that deletes all cuts made in a SlicerT instance.

https://github.com/user-attachments/assets/7bcb2539-cf52-43f4-9a0f-8a538929e0cd

bratpeki committed:
Initial, fixed

bratpeki committed:
setCheckable is unnecessary

bratpeki committed:
Connect and slot set up for the button

bratpeki committed:
I ACTUALLY DID IT???

bratpeki commented:
Trying to center the buttons right now, that's proving a bit difficult.
Current attempt looks like this:

![image](https://github.com/user-attachments/assets/24f87460-96c8-46c9-bba9-ef8a5f37f1ce)

I'll have to check the bounding box size, since I don't know if some row of pixels belongs to that box or not.
For now, I have this approach I won't be pushing; gotta do more digging:


```diff
-       m_syncToggle->move((width() - 100), m_y1 + 5);
-       m_clearButton->move((width() - 100), m_y1 + 30);
+       m_syncToggle->move((width() - 100), m_y1 - 7);
+       m_clearButton->move((width() - 100), m_y1 + 18);
```

bratpeki commented:
I think this looks pretty good:

![image](https://github.com/user-attachments/assets/4109dc8a-15b1-41d0-ab2e-3b94f560fdba)
![image](https://github.com/user-attachments/assets/6d295680-226d-4009-beaf-1e5d1b8455ba)

![image](https://github.com/user-attachments/assets/cb2b5019-7c23-4c29-a876-8111a173aee5)

I worked off the logic of `original offset of the first button` + `relative offset the second button begins at` - `half of the offset of the second button`. This gave me the numbers:

```cpp
m_syncToggle->move((width() - 100),  m_y1 + 5      - 11);
m_clearButton->move((width() - 100), m_y1 + 5 + 22 - 11);
```

which is just:

```cpp
m_syncToggle->move((width() - 100), m_y1 - 6);
m_clearButton->move((width() - 100), m_y1 + 16);
```

Literally just magic numbers! :rofl:

---

Pushing that for now, I might find some better numbers.

bratpeki committed:
Magic numbers 1

bratpeki committed:
Accidentally left a printf in

bratpeki committed:
Changed everything to use the term 'slice'. It's Slicer-T!

bratpeki committed:
clearAll switched to clearSlices

bratpeki committed:
tr() added

bratpeki committed:
1 pixel is better

AW1534 commented:
It freezes upon clicking the MIDI button after clearing all notes. this is consistently reproducible.


https://github.com/user-attachments/assets/577b3d34-0f2b-4a3e-96e0-5806ef5c5899

AW1534 commented on plugins/SlicerT/SlicerTView.cpp:
```diff
@@ -111,6 +117,12 @@ Knob* SlicerTView::createStyledKnob()
 
+// Clear all notes
+void SlicerTView::clearSlices() {
+	m_slicerTParent->m_slicePoints.clear();
```
```suggestion
	m_slicerTParent->m_slicePoints.clear();
	m_slicerTParent->m_slicePoints.push_back(0); // add slice at start of sample
```

AW1534 reviewed (changes requested):
https://github.com/user-attachments/assets/9248a883-4e71-49de-84a7-c85897cd5f9a

This fixes it

regulus79 commented on plugins/SlicerT/clearSlices_active.png:
Do we normally use camelCase for image file names?

AW1534 replied:
nope, usually snake_case

bratpeki commented:
Will adress both in the next commit 🫡

bratpeki commented:
Fixed! The changes include:

- Added a `push_back` for both `0` and `1`
- Changed the new image files' names to use snake case
- Removed the `m_clearAll` variable from the `SlicerT` class, since that doesn't have to be saved to a project

AW1534 commented:
I think you forgot to push 😆

bratpeki commented:
Doing final checks and pushing in a minute, I promise! :rofl:

bratpeki committed:
The long awaited!

bratpeki commented:
Please note that I'm still using a series of pluses and minuses to calculate buttons positions. I'd like to shorten that down before the merge. If this looks good, I'll shorten it right now:

![image](https://github.com/user-attachments/assets/caea8604-b0b6-4828-b772-3a7879c4d0cc)

AW1534 commented:
> Please note that I'm still using a series of pluses and minuses to calculate buttons positions. I'd like to shorten that down before the merge. If this looks good, I'll shorten it right now:

Looks good to me. You should add a comment that directs to this PR, which explains the mysterious values.

bratpeki committed:
Added comments

bratpeki commented:
@AW1534 Not a pushed change, just thinking out loud:

```diff
- brush.drawText(s_x4 - 8, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("Midi"));
+ brush.drawText(s_x4 - 8, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("MIDI")); // tr(...)? I'm pretty sure it's "MIDI" in every language lol
```

AW1534 commented:
> @AW1534 Not a pushed change, just thinking out loud:
> 
> ```diff
> - brush.drawText(s_x4 - 8, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("Midi"));
> + brush.drawText(s_x4 - 8, y1_text, s_textBoxWidth, s_textBoxHeight, Qt::AlignCenter, tr("MIDI")); // tr(...)? I'm pretty sure it's "MIDI" in every language lol
> ```

Yeah there's no need to translate MIDI.

bratpeki committed:
Appeased CodeFactor

AW1534 commented:
Anything blocking this? It's good for merge in my eyes

sakertooth commented on plugins/SlicerT/SlicerTView.cpp:
```diff
@@ -111,6 +117,19 @@ Knob* SlicerTView::createStyledKnob()
+
+	// Points are added to the start (0) and end (1) of the sample,
+	// so the whole sample can still be copied using MIDI.
+	m_slicerTParent->m_slicePoints.push_back(0);
+	m_slicerTParent->m_slicePoints.push_back(1);
```
You should generally prefer using `emplace_back` instead of `push_back`.

bratpeki replied:
Just read up on it, makes sense! Want me to replace the two calls?

sakertooth replied:
> Just read up on it, makes sense! Want me to replace the two calls?

Yes. Refer to [here](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-emplace.html) for an explanation as to why this is the case.

sakertooth commented:
Renaming the title to be more "actionable".

sakertooth commented:
> Literally just magic numbers! 🤣

We wanted them to use Qt layouts.. I think it was me and MG arguing for the use of layouts so stuff like this wouldn't be necessary, you would just add the widget and tada. However, the author preferred not do that, something along the lines of it was simpler not to use their system.

bratpeki committed:
emplace_back

bratpeki commented:
I think this is done. @AW1534 tested it already. @sakertooth please take a final look and, if all is well, merge.
